### PR TITLE
[SPARK-24816][SQL] SQL interface support repartitionByRange

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

SQL interface support `repartitionByRange` to improvement data pushdown.

## How was this patch tested?

unit tests, manual tests and benchmark tests.

I have test this feature with a big table(data size: 1.1 T, row count: 282,001,954,428).
The test SQL:
```sql
select * from table where id=401564838907
```
The test result:

Mode | Input Size | Records | Total Time | Duration | Prepare data Resource Allocation   MB-seconds
-- | -- | -- | -- | -- | --
default | 959.2 GB | 237624395522 | 11.2 h | 1.3 min | 6496280086
DISTRIBUTE BY | 970.8 GB | 244642791213 | 11.4 h | 1.3 min | 10536069846
SORT BY | 456.3 GB | 101587838784 | 5.4 h | 31 s | 8965158620
DISTRIBUTE BY + SORT BY | 219.0 GB | 51723521593 | 3.3 h | 54 s | 12552656774
RANGE PARTITION BY | 38.5 GB | 75355144 | 45 min | 13 s | 14525275297
RANGE PARTITION BY + SORT BY | 17.4 GB | 14334724 | 45 min | 12 s | 16255296698

It‘s obvious that after the data repartitioned by `RANGE PARTITION BY id SORT BY id`, only need to read 17.4 GB data can filtered `id=401564838907`, and the performance is greatly improved.

### CODE REVIEW

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -340,6 +340,7 @@ queryOrganization
     : (ORDER BY order+=sortItem (',' order+=sortItem)*)?
       (CLUSTER BY clusterBy+=expression (',' clusterBy+=expression)*)?
       (DISTRIBUTE BY distributeBy+=expression (',' distributeBy+=expression)*)?
+      (RANGE PARTITION BY rangePartitionBy+=expression (',' rangePartitionBy+=expression)*)?
```
Comment (wangyum) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: `RANGE DISTRIBUTE BY` and  `DISTRIBUTE BY` will conflict, so  used `RANGE PARTITION BY` here.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -2813,4 +2815,87 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       checkAnswer(df, Seq(Row(3, 99, 1)))
     }
   }
+
+  test("SQL interface support repartitionByRange") {
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Test suit copy from: https://github.com/apache/spark/blob/dfd7ac9887f89b9b51b7b143ab54d01f11cfcdb5/sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala#L379-L417

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 3]
```diff
@@ -306,44 +306,60 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
...
       // CLUSTER BY ...
       val expressions = expressionList(clusterBy)
       Sort(
         expressions.map(SortOrder(_, Ascending)),
         global = false,
         withRepartitionByExpression(ctx, expressions, query))
-    } else if (order.isEmpty && sort.isEmpty && distributeBy.isEmpty && clusterBy.isEmpty) {
+    } else if (!isOrder && !isSort && !isDistributeBy && !isClusterBy && isRangePartitionBy) {
+      // RANGE PARTITION BY ...
+      withRepartitionByExpression(ctx, rangePartitionBy.asScala.map(visitSortItem), query)
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Will invoke https://github.com/apache/spark/blob/e3201e165e41f076ec72175af246d12c0da529cf/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala#L888

**Comment by cloud-fan** [comment]
This is a PR to add a new SQL syntax for partitionByRange, which already have a DataFrame version. We should focus on the API itself instead of performance. Is there a similar syntax for repartitionByRange in other systems?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala [hunk 4]
```diff
@@ -394,6 +394,41 @@ class FilterPushdownBenchmark extends SparkFunSuite with BenchmarkBeforeAndAfter
       }
     }
   }
+
+  ignore("Pushdown benchmark for RANGE PARTITION BY/DISTRIBUTE BY") {
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala]: let's do it in another PR, this is not related to the API proposal.

**Comment by wangyum** [comment]
Teradata has `PARTITION BY RANGE_N` option in CREATE TABLE SQL statement. But SELECT SQL statement not seen in other systems.
https://info.teradata.com/HTMLPubs/DB_TTU_16_00/index.html#page/SQL_Reference/B035-1144-160K/nkt1472241480392.html

Or we use it as an enhanced version of ORDER BY.

**Comment by cloud-fan** [comment]
I don't have a good idea, cc @gatorsmile

**Comment by cloud-fan** [comment]
Maybe we can follow https://github.com/apache/spark/pull/21911 and use hint?

**Comment by gatorsmile** [comment]
+1

**Pull request closed without merging**